tests/test_pickling.cpp

```diff
@@ -60,6 +63,47 @@ void wrap(py::module m) {
 
 } // namespace exercise_trampoline
 
+namespace exercise_getinitargs_getstate_setstate {
+
+// Mechanism similar to what was established with Boost.Python:
+//   https://www.boost.org/doc/libs/1_58_0/libs/python/doc/v2/pickle.html
+// This mechanism was also adopted for PyCLIF:
+//   https://github.com/google/clif/blob/7d388e1de7db5beeb3d7429c18a2776d8188f44f/clif/testing/python/pickle_compatibility.clif
+
+class StoreTwoWithState {
+public:
+    StoreTwoWithState(int v0, int v1) : values_{v0, v1}, state_{"blank"} {}
+    const std::vector<int> &GetInitArgs() const { return values_; }
+    const std::string &GetState() const { return state_; }
+    void SetState(const std::string &state) { state_ = state; }
+
+private:
+    std::vector<int> values_;
+    std::string state_;
+};
+
+void wrap(py::module m) {
+    py::class_<StoreTwoWithState>(std::move(m), "StoreTwoWithState")
+        .def(py::init<int, int>())
+        .def("__getinitargs__",
+             [](const StoreTwoWithState &self) { return py::tuple(py::cast(self.GetInitArgs())); })
+        .def("__getstate__", &StoreTwoWithState::GetState)
+        .def(
```

Comment:
I am a bit confused that both reduce_ex and setstate are provided in the same python class. I thought this would completely bypass setstate during unpickling?

Reply:
I added this to the comment further up:

```
// The code below exercises the pickle protocol intentionally exactly as used
// in PyCLIF, to ensure that pybind11 stays fully compatible with existing
// client code relying on the long-established protocol. (It is impractical to
// change all such client code.)
```

The `__reduce_ex__` does indeed not necessarily have to call `__getinitargs__` and `__getstate__`, but I feel reducing the protocol as used in the wild just for the test would be a loss.

BTW: `__reduce__` (Boost.Python) / `__reduce_ex__` (PyCLIF) is the trick to make pickling robust for any Python version and pickle protocol. In theory `__getinitargs__` and `__getstate__` are not needed are could be replaced by another protocol, but in 2002 I decided to emulate what I found at that time in the Python documentation (for native types).

Reply:
I see. This is to make a class that supports all different versions of the pickle protocol. My Python experience only started around 2010, so I do lack that part of background.